src/ert/scheduler/scheduler.py

```diff
@@ -83,8 +76,31 @@ def kill_all_jobs(self) -> None:
         for task in self._tasks.values():
             task.cancel()
 
-    def stop_long_running_jobs(self, minimum_required_realizations: int) -> None:
-        pass
+    async def _stop_long_running_jobs(self, minimum_required_realizations: int) -> None:
+        LONG_RUNNING_FACTOR = 1.25
+        while True:
+            completed_jobs = []
+            for job_id, job in self._jobs.items():
+                try:
+                    if job.returncode.result() != 0:
+                        continue
+                    completed_jobs.append(job_id)
+                except (asyncio.CancelledError, asyncio.InvalidStateError):
+                    continue
+
+            if len(completed_jobs) >= minimum_required_realizations:
+                average_runtime = sum(
+                    self._jobs[job_id].running_duration for job_id in completed_jobs
+                ) / len(completed_jobs)
+                for job_id, task in self._tasks.items():
+                    if (
+                        self._jobs[job_id].running_duration
+                        > LONG_RUNNING_FACTOR * average_runtime
+                        and not task.done()
+                    ):
+                        task.cancel()
+                        await task
+            await asyncio.sleep(0.1)
```

Comment:
Perhaps this whole mechanism should be handled by `_process_event_queue()`? This task loops through all realisations multiple times, every 0.1 seconds.

The only time something can happen is when a job starts or finishes (either fail or success). Otherwise, if no job has finished since last loop, all jobs' `running_duration` have increased by `0.1`, but `average_runtime` stays the same. So once `len(completed_jobs) >= minimum_required_realizations` is `True`, ERT will keep on running until the jobs get 90% done and _then_ it kills them, and keep checking every 0.1s even though the time when we should start killing jobs is entirely predictable.

Instead, `_process_event_queue()` could have this logic that is only executed when a job starts or stops. This would mean no more checking every 100ms, which is nice imo because performance.

Reply:
I agree and makes sense, which brings to question why it is not so in old jobqueue - from where I got inspired? :)